nexus/reconfigurator/planning/src/blueprint_builder/external_networking.rs

```diff
@@ -295,9 +304,218 @@ impl<T: Hash + Eq> Iterator for AvailableIterator<'_, T> {
     }
 }
 
+// External IPs come in two flavors, from an allocation point of view: IPs that
+// require exclusive use, and SNAT IPs that can be shared amongst up to four
+// services, because the port range is broken up into four 16384-sized chunks.
+// This struct keeps track of both kinds of IPs used by blueprints, allowing
+// allocation of either kind.
+#[derive(Debug)]
+struct ExternalIpAllocator<'a> {
+    service_ip_pool_ips: DebugIgnore<Box<dyn Iterator<Item = IpAddr> + 'a>>,
+    used_exclusive_ips: BTreeSet<IpAddr>,
+    used_snat_ips: BTreeMap<IpAddr, BTreeSet<SnatPortRange>>,
+}
+
+impl<'a> ExternalIpAllocator<'a> {
+    fn new(service_pool_ranges: &'a [IpRange]) -> Self {
+        let service_ip_pool_ips =
+            service_pool_ranges.iter().flat_map(|r| r.iter());
+        Self {
+            service_ip_pool_ips: DebugIgnore(Box::new(service_ip_pool_ips)),
+            used_exclusive_ips: BTreeSet::new(),
+            used_snat_ips: BTreeMap::new(),
+        }
+    }
+
+    fn mark_ip_used(
+        &mut self,
+        external_ip: &OmicronZoneExternalIp,
+    ) -> anyhow::Result<()> {
+        match external_ip {
+            OmicronZoneExternalIp::Floating(ip) => {
+                let ip = ip.ip;
+                if self.used_snat_ips.contains_key(&ip)
+                    || !self.used_exclusive_ips.insert(ip)
+                {
+                    bail!("duplicate external IP: {external_ip:?}");
```

Comment:
Is this something we'll ever hit beyond a logic error?

IIRC there was an error enum in this code where internal logic errors could be put -- could this use that enum to be explicit?

Reply:
Ah if it's already the prevailing standard then this is okay. I'd say it's definitely worth doing at some point though, anyhow errors floating around just make me uncomfortable because of how untyped they are.